wasm/validator/src/default.rs

```diff
@@ -112,18 +112,30 @@ pub(crate) use map_all_crate_tokens;
 pub(crate) use tokens;
 
 impl Validate for DefaultValidator {
-    fn permission_tokens() -> Vec<PermissionTokenDefinition> {
-        let mut v = Vec::new();
+    fn migrate() -> MigrationResult {
+        let old_definitions = BTreeSet::from_iter(
+            FindAllPermissionTokenDefinitions
+                .execute()
+                .map_err(|error| {
+                    Self::format_err(error, "Failed to find all permission token definitions")
+                })?,
+        );
 
-        macro_rules! add_to_vec {
-            ($token_ty:ty) => {
-                v.push(<$token_ty as ::iroha_validator::permission::Token>::definition());
-            };
-        }
+        let new_definitions = {
+            let mut set = BTreeSet::new();
+
+            macro_rules! add_to_set {
+                ($token_ty:ty) => {
+                    set.insert(<$token_ty>::definition());
+                };
+            }
 
-        map_all_crate_tokens!(add_to_vec);
+            map_all_crate_tokens!(add_to_set);
+            set
+        };
 
-        v
+        Self::unregister_old_permission_tokens(&old_definitions, &new_definitions)?;
+        Self::register_permission_tokens(&old_definitions, &new_definitions)
```

Comment:
> we have to make sure that it's impossible to call any of migration functions (line unregister_old_permission_tokens) at any time except during migration

Actually this is already done by our `wasm::Runtime` design.

I think that *Validator/Executor* should be able to run custom hook during upgrade, which migration is. If we will make migration in two sides: something on Iroha side and something on Wasm side that will be too implicit and will require more work from both sides

Reply:
We had a relatively long discussion on this topic, so I think it's worth reiterating the key points. 

1. Our attack vector should exclude the Executor binary compromise. Not because it can't happen, but because it's _game over_ and nothing can be done to mitigate the consequences. 
2. The only constraint that the `migration` entrypoint must satisfy is that it is reproducible, and reasonable. What constitutes _reasonable_ very much depends on the blockchain, hence why the executor is the responsible party. Violating either constraint can be fixed by removing the migration from block history and starting over. 
3. If Iroha handles migrations, it must be made aware of (and be taught how to enforce the) executor invariants. This means that code is (as @Arjentix mentioned) shared between Iroha and the executor. 

Under these circumstances, it might be worth exploring both ideas, but it seems that a lot of the advantages of host-side migrations are compromised by points 1, 2 and 3. IMO we'll have a smaller API surface area if we allow the executor, an already important piece of the blockchain security, to also be responsible for its own migrations.

Reply:
After a discussion with @appetrosyan and @mversic we decided to keep following current approach AFAIU.

@mversic , if you're agree, close the discussion, please